src/drivers/imx/dummy-dma.c

```diff
@@ -45,54 +134,302 @@ static int dummy_dma_status(struct dma_chan_data *channel,
+			 */
+			bytes += size;
+			break;
+		}
+
+		/* Confirm transfer for now, adding it up */
+		bytes -= size;
+		if (bytes <= 0) /* Transfer is done */
+			break;
+	}
+
+	/* How many bytes did we copy in total? */
+	return orig_bytes - bytes;
+}
+
+/* Since there isn't any hardware we must actually perform the copies
+ * synchronously within this function, then let the host component know
+ * how much data we've got.
+ */
 static int dummy_dma_copy(struct dma_chan_data *channel, int bytes,
 			  uint32_t flags)
 {
+	struct dma_cb_data next;
+
+	next.elem.size = docopies(channel, bytes);
+
+	/* Call the callback to announce how much we copied */
+	if (channel->cb_type & DMA_CB_TYPE_COPY)
+		channel->cb(channel->cb_data, DMA_CB_TYPE_COPY, &next);
+
 	return 0;
 }
 
 static int dummy_dma_probe(struct dma *dma)
 {
+	struct dma_chan_data *dma_chans;
+	struct dma_chan_pdata *chanp;
+	int i;
+
+	if (dma_get_drvdata(dma)) {
+		trace_dummydma_error("dummy-dmac %d already created!",
+				     dma->plat_data.id);
+		return -EEXIST; /* already created */
+	}
+
+	dma_chans = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			    SOF_MEM_CAPS_RAM,
+			    dma->plat_data.channels * sizeof(dma_chans[0]));
+	if (!dma_chans) {
+		trace_dummydma_error("dummy-dmac %d: Out of memory!",
+				     dma->plat_data.id);
+		return -ENOMEM;
+	}
+
+	chanp = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			SOF_MEM_CAPS_RAM,
+			dma->plat_data.channels * sizeof(chanp[0]));
+	if (!chanp) {
+		rfree(dma_chans);
+		trace_dummydma_error("dummy-dmac %d: Out of memory!",
+				     dma->plat_data.id);
+		return -ENOMEM;
+	}
+
+	dma_set_drvdata(dma, dma_chans);
+
+	for (i = 0; i < dma->plat_data.channels; i++) {
+		dma_chans[i].dma = dma;
+		dma_chans[i].index = i;
+		dma_chans[i].status = COMP_STATE_INIT;
+		dma_chans[i].private = &chanp[i];
+	}
+
+	atomic_init(&dma->num_channels_busy, 0);
+
 	return 0;
 }
 
 static int dummy_dma_remove(struct dma *dma)
 {
+	struct dma_chan_data *chans = dma_get_drvdata(dma);
+
+	tracev_dummydma("dummy_dma %d -> remove", dma->plat_data.id);
+	if (!chans)
+		return 0;
+
+	rfree(dma_chan_get_data(&chans[0]));
+	rfree(chans);
+	dma_set_drvdata(dma, NULL);
 	return 0;
 }
 
 static int dummy_dma_get_data_size(struct dma_chan_data *channel,
 				   uint32_t *avail, uint32_t *free)
 {
+	/* TODO remove hardcode, if possible */
+	switch (channel->direction) {
+	case DMA_DIR_HMEM_TO_LMEM:
+		*avail = 384;
+		break;
+	case DMA_DIR_LMEM_TO_HMEM:
+		*free = 384;
+		break;
+	default:
+		trace_dummydma_error("get_data_size direction: %d",
+				     channel->direction);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int dummy_dma_get_attribute(struct dma *dma, uint32_t type,
+				   uint32_t *value)
+{
+	switch (type) {
+	case DMA_ATTR_BUFFER_ALIGNMENT:
+	case DMA_ATTR_COPY_ALIGNMENT:
+		*value = 4; /* Align to uint32_t; not sure, maybe 1 works? */
+		return 0;
+	default:
+		return -ENOENT; /* Attribute not found */
+	}
 	return 0;
 }
 
```

Comment:
I think a good approach would be also to split this driver implementation into multiple patches.

1 patch for related functions. e.g probe / remove etc.

Reply:
To make it easier to review? Because from the functionality point of view it will only work with the entire patch added (all functions).